clusterloader2/pkg/measurement/util/runtimeobjects/replicaswatcher.go

```diff
@@ -0,0 +1,159 @@
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util/informer"
+)
+
+const (
+	informerSyncTimeout = time.Minute
+)
+
+// ReplicasWatcher is a struct that allows to check a number of replicas at a given time.
+// Usage:
+// var rw ReplicasWatcher = (...)
+// if err := rw.Start(stopCh); err != nil {
+//   panic(err);
+// }
+// // Get number of replicas as needed.
+// val = rw.Replicas()
+// ...
+// val = rw.Replicas()
+type ReplicasWatcher interface {
+	Replicas() int
+	// Start must block until Replicas() returns a correect value.
+	Start(stopCh chan struct{}) error
+}
+
+// ConstReplicas is a ReplicasWatcher implementation that returns a constant value.
+type ConstReplicas struct {
+	Const int
+}
+
+func (c *ConstReplicas) Replicas() int {
+	return c.Const
+}
+
+func (c *ConstReplicas) Start(_ chan struct{}) error {
+	return nil
+}
+
+var _ ReplicasWatcher = &ConstReplicas{}
+
+// NodeCounter counts a number of node objects matching nodeSelector and affinity.
+type NodeCounter struct {
+	client       clientset.Interface
+	nodeSelector labels.Selector
+	affinity     *corev1.Affinity
+
+	mu       sync.Mutex
+	replicas int
+}
+
+var _ ReplicasWatcher = &NodeCounter{}
+
+// NewNodeCounter returns a new nodeCounter that return a number of objects matching nodeSelector and affinity.
+func NewNodeCounter(client clientset.Interface, nodeSelector labels.Selector, affinity *corev1.Affinity) *NodeCounter {
+	return &NodeCounter{
+		client:       client,
+		nodeSelector: nodeSelector,
+		affinity:     affinity,
+	}
+}
+
+func (n *NodeCounter) Start(stopCh chan struct{}) error {
+	i := informer.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = n.nodeSelector.String()
+				return n.client.CoreV1().Nodes().List(context.TODO(), options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = n.nodeSelector.String()
+				return n.client.CoreV1().Nodes().Watch(context.TODO(), options)
+			},
+		},
+		func(oldObj, newObj interface{}) {
+			if err := n.handleObject(oldObj, newObj); err != nil {
+				klog.Errorf("Error while processing node: %v", err)
+			}
+		},
+	)
+	// StartAndSync blocks until elements from initial list call are processed.
+	return informer.StartAndSync(i, stopCh, informerSyncTimeout)
+}
+
+func (n *NodeCounter) handleObject(oldObj, newObj interface{}) error {
+	old, err := n.shouldRun(oldObj)
+	if err != nil {
+		return err
+	}
+	new, err := n.shouldRun(newObj)
+	if err != nil {
+		return err
+	}
+	if new == old {
+		return nil
+	}
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if new && !old {
```

Comment:
nit: feel fre to ignore:

n.replicas += (new == true)
n.replicas -= (old == true)

[then you can skip the if new == old too]

Reply:
I was thinking on something like "n.repliacas += int(new) - int(old)" already, but in golang it's not possible to easily cast bool -> int.
I can change return type for "shouldRun" to int, but it will require implementing some if inside, so it's not clear if this is overall win.

See https://play.golang.org/p/qurKupUdp7b